src/reconcile/rec_visibility.c

```diff
@@ -868,27 +983,71 @@ __rec_fill_tw_from_upd_select(
      * that the value is visible to any timestamp/transaction id ahead of it.
      */
+    bool write_start_prepare = write_prepare;
     if (upd->type == WT_UPDATE_TOMBSTONE) {
-        WT_TIME_WINDOW_SET_STOP(select_tw, upd, upd->prepare_state);
+        WT_TIME_WINDOW_SET_STOP(select_tw, upd, write_prepare);
         tombstone = upd_select->tombstone = upd;
+        WT_ACQUIRE_READ(tombstone_txnid, tombstone->txnid);
+        if (tombstone_txnid == WT_TXN_ABORTED)
+            tombstone_txnid = tombstone->upd_saved_txnid;
+        /*
+         * Find the update this tombstone applies to.
+         *
+         * We need to find the full update if the prepared tombstone is rolled back. We resolve
+         * prepare updates recursively from the oldest to the newest. If we write a prepared
+         * tombstone, we may see a prepared update that is rolled back from the same transaction.
+         * Ensure we also write it to the disk in this case.
+         */
+        if (write_prepare || !__wt_txn_upd_visible_all(session, upd)) {
+            uint64_t next_txnid = WT_TXN_NONE;
+            for (; upd->next != NULL; upd = upd->next) {
+                WT_ACQUIRE_READ(next_txnid, upd->next->txnid);
+                write_start_prepare = write_prepare && next_txnid == tombstone_txnid;
+                if (next_txnid != WT_TXN_ABORTED)
+                    break;
 
-        /* Find the update this tombstone applies to. */
-        if (!__wt_txn_upd_visible_all(session, upd)) {
-            while (upd->next != NULL && upd->next->txnid == WT_TXN_ABORTED)
-                upd = upd->next;
+                if (!write_start_prepare)
+                    continue;
 
-            WT_ASSERT(session, upd->next == NULL || upd->next->txnid != WT_TXN_ABORTED);
+                if (!F_ISSET(S2C(session), WT_CONN_PRESERVE_PREPARED))
+                    continue;
+
+                /* We may see aborted reserve updates in between the prepared updates. */
+                if (upd->next->type == WT_UPDATE_RESERVE)
+                    continue;
+
+                if (upd->next->prepare_state != WT_PREPARE_INPROGRESS) {
+                    write_start_prepare = false;
+                    continue;
+                }
+
+                /*
+                 * Since we resolve prepared update from the oldest to newest, we may see a
+                 * tombstone still in prepared state but a prepared update that is rolled back from
+                 * the same transaction here. Handle this case.
+                 */
+                if (upd->next->upd_saved_txnid == tombstone_txnid) {
+                    WT_ASSERT(session, upd->next->prepare_ts == tombstone->prepare_ts);
+                    break;
+                } else {
+                    write_start_prepare = false;
+                }
+            }
+            WT_ASSERT(session,
+              upd->next == NULL || next_txnid != WT_TXN_ABORTED ||
```

Comment:
May be this is fine. If we race with commit, next_txnid should not be aborted so we return true. But the logic is a bit difficult to understand. There are many different race cases. We can add a comment discuss each case.

Reply:
updated the comment and assert